### ISSUE TITLE
Extract training or validation losses for use in another function

### ISSUE THREAD

**Issue by ayindemalik** [ultralytics/yolov5#9317]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello, I am working on a project related to handwritten detection. apart from normal training, I need to perform also adversarial training using adversarial samples. However, to generate adversarial samples, training or validation losses are needed. So my issue is how I could extract or save the losses during the training process and be able to use them in generating adversarial images using functions like FGSM (Fast Gradient Sign Method) or PGD (Projected Gradient Descent).

Any help is appreciated.

**Comment by glenn-jocher** [comment]
Losses are computed in train.py here:
https://github.com/ultralytics/yolov5/blob/5a134e06530a8c24fdb9774c2c4ab0b513b08260/train.py#L307

**Comment by ayinemalik** [comment]
thanks Mr. glenn-jocker for your helped reply. I still need an additional htlp tip

sopose I want to use the yolov5 model during the training to generate adversarial samples images and include them through the training process like it has been processed in the code picture below for image classification  model.

So question is at with which process and line of code I could get the model and apply it to **fast_gradient_method(torch_model, xs, eps=0.3, norm=np.inf, clip_min=0., clip_max=1.)**
method in yolov5. or any additional tips will be apreciated.
![image](https://user-images.githubusercontent.com/27240587/189134848-bc414c68-1e12-4ce5-b31b-be7f0808729b.png)
 

Thanks in advance.

**Comment by glenn-jocher** [comment]
@ayindemalik you can apply adversarial attacks directly to the YOLOv5 model by accessing the underlying PyTorch model. In train.py, after line 307, insert the following code to generate adversarial samples:

```python
import torch
import numpy as np
from advertorch.attacks import GradientSignAttack

# Access PyTorch model from YOLOv5 model
torch_model = model.model[-1]  # Access the underlying PyTorch model

# Generate adversarial samples
adversary = GradientSignAttack(torch_model)
adversarial_samples = adversary.perturb(xs, ys)
```
This code snippet accesses the PyTorch model from YOLOv5 and generates adversarial samples using the FGSM method. Remember to adjust the attack method and parameters according to your requirements.
